pika/connection.py

```diff
@@ -947,66 +902,60 @@ def _set_url_socket_timeout(self, value):
+                        context=ssl_sock.context,
+                        server_hostname=ssl_sock.server_hostname)
+                finally:
+                    ssl_sock.close()
+            finally:
+                sock.close()
+
+
 
     def _set_url_tcp_options(self, value):
         """Deserialize and apply the corresponding query string arg"""
         self.tcp_options = ast.literal_eval(value)
 
+
 class SSLOptions(object):
     """Class used to provide parameters for optional fine grained control of SSL
     socket wrapping.
 
-    :param string keyfile: The key file to pass to SSLContext.load_cert_chain
-    :param string key_password: The key password to passed to
-                                                    SSLContext.load_cert_chain
-    :param string certfile: The certificate file to passed to
-                                                    SSLContext.load_cert_chain
-    :param bool server_side: Passed to SSLContext.wrap_socket
-    :param verify_mode: Passed to SSLContext.wrap_socket
-    :param ssl_version: Passed to SSLContext init, defines the ssl
-                                                                version to use
-    :param string cafile: The CA file passed to
-                                            SSLContext.load_verify_locations
-    :param string capath: The CA path passed to
-                                            SSLContext.load_verify_locations
-    :param string cadata: The CA data passed to
-                                            SSLContext.load_verify_locations
-    :param do_handshake_on_connect: Passed to SSLContext.wrap_socket
-    :param suppress_ragged_eofs: Passed to SSLContext.wrap_socket
-    :param ciphers: Passed to SSLContext.set_ciphers
-    :param server_hostname: SSLContext.wrap_socket, used to enable SNI
     """
 
-    def __init__(self,
-                 keyfile=None,
-                 key_password=None,
-                 certfile=None,
-                 server_side=False,
-                 verify_mode=ssl.CERT_NONE,
-                 ssl_version=ssl.PROTOCOL_SSLv23,
-                 cafile=None,
-                 capath=None,
-                 cadata=None,
-                 do_handshake_on_connect=True,
-                 suppress_ragged_eofs=True,
-                 ciphers=None,
-                 server_hostname=None):
-        self.keyfile = keyfile
-        self.key_password = key_password
-        self.certfile = certfile
-        self.server_side = server_side
-        self.verify_mode = verify_mode
-        self.ssl_version = ssl_version
-        self.cafile = cafile
-        self.capath = capath
-        self.cadata = cadata
-        self.do_handshake_on_connect = do_handshake_on_connect
-        self.suppress_ragged_eofs = suppress_ragged_eofs
-        self.ciphers = ciphers
+    # Protect against accidental assignment of an invalid attribute
+    __slots__ = ('context', 'server_hostname')
+
+    def __init__(self, context, server_hostname=None):
```

Comment:
We're keeping `SSLOptions` around just for `server_hostname`. We could get rid of this class and add `ssl_server_hostname` to `ConnectionParameters`. Just an idea.